pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_resource_controller.go

```diff
@@ -0,0 +1,205 @@
+	"strings"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	kcpclient "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	apisinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apis/v1alpha1"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	"github.com/kcp-dev/kcp/pkg/informer"
+)
+
+const (
+	controllerNameResource = "kcp-resource-permissionclaimlabel"
+)
+
+// NewController returns a new controller for APIBindings.
+func NewResourceController(
+	kcpClusterClient kcpclient.Interface,
+	dynamicClusterClient dynamic.Interface,
+	dynamicDiscoverySharedInformerFactory *informer.DynamicDiscoverySharedInformerFactory,
+	apiBindingInformer apisinformers.APIBindingInformer,
+) (*resourceController, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerNameResource)
+
+	c := &resourceController{
+		queue:                queue,
+		kcpClusterClient:     kcpClusterClient,
+		dynamicClusterClient: dynamicClusterClient,
+		ddsif:                dynamicDiscoverySharedInformerFactory,
+
+		apiBindingsLister: apiBindingInformer.Lister(),
+		listAPIBindings: func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error) {
+			obj, err := apiBindingInformer.Informer().GetIndexer().ByIndex(indexers.ByLogicalCluster, clusterName.String())
+			var ret []*apisv1alpha1.APIBinding
+			if err != nil {
+				return ret, err
+			}
+
+			for _, o := range obj {
+				binding, ok := o.(*apisv1alpha1.APIBinding)
+				if !ok {
+					klog.Errorf("unexpected type: %T, got %T", &apisv1alpha1.APIBinding{}, o)
+				}
+				ret = append(ret, binding)
+			}
+
+			return ret, nil
+		},
+		apiBindingsIndexer: apiBindingInformer.Informer().GetIndexer(),
+	}
+
+	c.ddsif.AddEventHandler(informer.GVREventHandlerFuncs{
+		AddFunc:    func(gvr schema.GroupVersionResource, obj interface{}) { c.enqueueForResource(gvr, obj) },
+		UpdateFunc: func(gvr schema.GroupVersionResource, _, obj interface{}) { c.enqueueForResource(gvr, obj) },
+		DeleteFunc: nil, // Nothing to do.
+	})
+
+	return c, nil
+
+}
+
+// resourceController reconciles resources from the ddsif, and will determine if it needs
+// its permission claim labels updated.
+type resourceController struct {
+	queue workqueue.RateLimitingInterface
+
+	kcpClusterClient     kcpclient.Interface
+	apiBindingsIndexer   cache.Indexer
+	dynamicClusterClient dynamic.Interface
+	ddsif                *informer.DynamicDiscoverySharedInformerFactory
+
+	apiBindingsLister apislisters.APIBindingLister
```

Comment:
(no action required) the pattern of creating function pointer fields such as `listAPIBindings` for easier testability can also apply to `getAPIBinding`. If you wanted to, you could have a `getAPIBinding` function pointer field instead of using the lister directly.